document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java

```diff
@@ -386,28 +387,58 @@ public long total(Query query) {
   @Override
   public boolean bulkUpsert(Map<Key, Document> documents) {
     try {
-      List<UpdateOneModel<BasicDBObject>> bulkCollection = new ArrayList<>();
-      for (Entry<Key, Document> entry : documents.entrySet()) {
-        Key key = entry.getKey();
-        // insert or overwrite
-        bulkCollection.add(new UpdateOneModel<>(
-            this.selectionCriteriaForKey(key),
-            prepareUpsert(key, entry.getValue()),
-            new UpdateOptions().upsert(true)));
-      }
-
-      BulkWriteResult result = Failsafe.with(bulkWriteRetryPolicy)
-          .get(() -> collection.bulkWrite(bulkCollection, new BulkWriteOptions().ordered(false)));
+      BulkWriteResult result = bulkUpsertImpl(documents);
       LOGGER.debug(result.toString());
-
       return true;
-
     } catch (IOException | MongoServerException e) {
       LOGGER.error("Error during bulk upsert for documents:{}", documents, e);
       return false;
     }
   }
 
+  private BulkWriteResult bulkUpsertImpl(Map<Key, Document> documents) throws JsonProcessingException {
+    List<UpdateOneModel<BasicDBObject>> bulkCollection = new ArrayList<>();
+    for (Entry<Key, Document> entry : documents.entrySet()) {
+      Key key = entry.getKey();
+      // insert or overwrite
+      bulkCollection.add(new UpdateOneModel<>(
+          this.selectionCriteriaForKey(key),
+          prepareUpsert(key, entry.getValue()),
+          new UpdateOptions().upsert(true)));
+    }
+
+    return Failsafe.with(bulkWriteRetryPolicy)
+        .get(() -> collection.bulkWrite(bulkCollection, new BulkWriteOptions().ordered(false)));
+  }
+
+  @Override
+  public Iterator<Document> bulkUpsertAndReturnOlderDocuments(Map<Key, Document> documents) throws IOException {
+    try {
+      // First get all the documents for the given keys.
+      FindIterable<BasicDBObject> cursor = collection.find(selectionCriteriaForKeys(documents.keySet()));
+      final MongoCursor<BasicDBObject> mongoCursor = cursor.cursor();
+
+      // Now go ahead and do the bulk upsert.
+      BulkWriteResult result = bulkUpsertImpl(documents);
+      LOGGER.debug(result.toString());
+
+      return new Iterator<>() {
+        @Override
+        public boolean hasNext() {
+          return mongoCursor.hasNext();
+        }
+
+        @Override
+        public Document next() {
+          return MongoCollection.this.dbObjectToDocument(mongoCursor.next());
+        }
+      };
+    } catch (JsonProcessingException e) {
+      LOGGER.error("Error during bulk upsert for documents:{}", documents, e);
```

Comment:
This is more like an in general comment. I am usually in favour of either log the error and handle it or bubble up the exception but not both of them because they usually flood the logs. Also, do we want to print the full set of documents in logs? How about privacy concerns and also efficient usage of the log storage? I don't thing dumping the failing documents in the logs is actionable either.

Reply:
Raised https://github.com/hypertrace/document-store/pull/25